### ISSUE TITLE
iOS 12 support

### ISSUE THREAD

**Issue by gabrc52** [iandwelker/smserver#129]
Opening to see if someone has the skills for supporting iOS 12 (and maybe lower versions)? 

Alternatively, the UI be decoupled from the server, if it works and there's no UI or only a command-line tool where you have to open a Terminal on the iPhone or SSH to it in order to configure it, it's still pretty fine.

> Do you plan support for IOS 12?

> Unfortunately, the app is written in SwiftUI, which only supports iOS 13+, and I don’t know UIKit (which is what iOS 12 uses) Its all open source, though, and I would totally welcome another developer forking it and adding support for iOS 12 and below. 
https://www.reddit.com/r/jailbreak/comments/i7v04g/free_release_smserver_send_texts_from_your/g14cbpg?utm_source=share&utm_medium=web2x&context=3

**Comment by iandwelker** [comment]
Ok, so turns out it was crazily easy to make a (potentially) iOS 12-compatible CLI version of SMServer ([commit](https://github.com/iandwelker/smserver/commit/d418980dbcf3ce3716c4d154ff860094b430a81d)). 

I don't have an iOS 12 device to test with to verify, but it runs fine (mind you, CLI only &mdash; the app is just completely blank if you launch it) on my iOS 14.4.2 6s. I'll upload a build of the 12-compatible version here in a minute for others to test out and verify.

**Comment by CSab6482** [comment]
![image](https://user-images.githubusercontent.com/53708281/119170290-0d0c3600-ba31-11eb-8cb8-f520b13b8fee.png)
I was unable to download the .deb, can you upload it to a website like MediaFire, Google Drive, or DropBox?

**Comment by iandwelker** [comment]
@CSab6482  sorry about that -- I just fixed [the link](https://byu.box.com/v/smserver12-080), and updated the make script as well so that it will work on your mac. Unfortunately, there's no way (as of right now) to build it on windows.

**Comment by CSab6482** [comment]
I got it installed on my iPod, and just like you said, the app just shows a black screen when I open it. How do I start the server? Through terminal/SSH commands?
![image](https://user-images.githubusercontent.com/53708281/119186105-9f1e3980-ba45-11eb-8c3b-5f0f6b7effc8.png)
![IMG_0001](https://user-images.githubusercontent.com/53708281/119186405-0fc55600-ba46-11eb-8a46-05fc630e48af.PNG)

**Comment by iandwelker** [comment]
Yup, SSH in and try running `/Applications/SMServer12/SMServer12 -r` to see if it will show that the server has started up

**Comment by CSab6482** [comment]
![image](https://user-images.githubusercontent.com/53708281/119187514-74cd7b80-ba47-11eb-95dd-8117a5f662f2.png)
I tried SMServer12 -r, SMServer12.app -r, and SMServer -r. Am I in the wrong directory or running the wrong commmand?

**Comment by CSab6482** [comment]
Another update, when I open Sileo on my iPod, I now receive this error message, after which the app stays frozen. I am going to Restore RootFS using checkra1n and rejailbreak. Maybe an error occurred during my installation of the .deb? I'll retry this with Cydia/Substitute instead of Sileo/Procursus to see if it makes a difference.
![Image-1](https://user-images.githubusercontent.com/53708281/119188091-2ff61480-ba48-11eb-9d09-edabb765ba01.jpg)

**Comment by gabrc52** [comment]
> I tried SMServer12 -r, SMServer12.app -r, and SMServer -r. Am I in the wrong directory or running the wrong commmand?

@CSab6482 Try ./SMServer12 

Unlike Windows, the current directory is not part of the PATH. The dot means current directory.

As for the DPKG error try running `sudo dpkg --reconfigure -a`

**Comment by CSab6482** [comment]
When I have Cydia set as my package manager, I received an error. However, the SMServer app still showed up after I ran uicache
![image](https://user-images.githubusercontent.com/53708281/119192488-3c7d6b80-ba4e-11eb-8a56-7936fafed24c.png)
Cydia then gave me this error message
![Image-1](https://user-images.githubusercontent.com/53708281/119192714-90885000-ba4e-11eb-8ad0-b3b6dcfbc9b5.jpg)

**Comment by CSab6482** [comment]
> > I tried SMServer12 -r, SMServer12.app -r, and SMServer -r. Am I in the wrong directory or running the wrong commmand?
> 
> @CSab6482 Try ./SMServer12
> 
> Unlike Windows, the current directory is not part of the PATH. The dot means current directory.
> 
> As for the DPKG error try running `dpkg --reconfigure -a`

This worked! 
![image](https://user-images.githubusercontent.com/53708281/119192912-d5ac8200-ba4e-11eb-85d1-ea280e94b5c2.png)
I'll see if the `dpkg --reconfigure -a` command fixes my problems with Cydia and Sileo, and I'll put my findings in a comment.

**Comment by CSab6482** [comment]
I attempted to send a message and I received this error. I'll install cr4shed and upload any error logs here.
![SMServer12](https://user-images.githubusercontent.com/53708281/119193804-325c6c80-ba50-11eb-99d5-f708f3b6806f.png)

**Comment by CSab6482** [comment]
> > I tried SMServer12 -r, SMServer12.app -r, and SMServer -r. Am I in the wrong directory or running the wrong commmand?
> 
> @CSab6482 Try ./SMServer12
> 
> Unlike Windows, the current directory is not part of the PATH. The dot means current directory.
> 
> As for the DPKG error try running `dpkg --reconfigure -a`

I was not able to run `dpkg --reconfigure -a` while using Sileo/Procursus
![image](https://user-images.githubusercontent.com/53708281/119195014-ee6a6700-ba51-11eb-80bd-9d24df49c6f7.png)

**Comment by CSab6482** [comment]
As far as I can tell, it works perfectly on iOS 12. The only error I could find was that installing the .deb broke my package managers, but I'm sure that once the tweak is available on Twickd, this problem will not occur.
![SMServer 12 5 3](https://user-images.githubusercontent.com/53708281/119195953-700ec480-ba53-11eb-91d9-15f092e24ccc.png)

**Comment by iandwelker** [comment]
> Does this confirm that SMServer will also work on iOS 10 or 11? Or do we need other people to test that?

I think we'll need to test those as well. I don't know if I'm going to release this on a public repository (since it's CLI-only (for now, at least), and that would confuse people), so it's not crucial that I do extensive testing for those versions, but I think I'll upload a compiled version here, with instructions on how to use, if people would like to use it.

**Comment by stevenmirabito** [comment]
I successfully installed the deb posted earlier on an old iPhone 5c with iOS 10.3.3 and can run the server. Some notes:
- iOS 10.3.3 does not support LZMA compression out of the box, so I needed to install the "XZ Utils" tweak from Cydia for dpkg to unpack the deb
- Specific to the iPhone 5c (32-bit so no APFS and very little space on the system partition), I needed to use iCleaner to remove languages and keyboards to make enough room to install SMServer
- The `dpkg` errors appear to stem from the `postinst` script checking for `/Applications/SMServer.app/SMServer` whereas the binary for this build unpacks to `/Applications/SMServer12.app/SMServer12`

However, it appears the bundled `libsmserver` is 64-bit only? The UI doesn't load any messages nor can I send messages, and I see this in the syslog:

```
Jun 13 19:55:57 iMessage-Server syncdefaultsd[1015] <Notice>: MS:Notice: Injecting: com.apple.syncdefaultsd [syncdefaultsd] (1349.70)
Jun 13 19:55:57 iMessage-Server syncdefaultsd[1015] <Error>: MS:Error: binary does not support this cpu type
Jun 13 19:55:57 iMessage-Server syncdefaultsd[1015] <Error>: MS:Error: failure to check libsmserver.dylib
```

**Comment by iandwelker** [comment]
@stevenmirabito  thank you so much for checking that out! With regard to the issues you listed:
- I'll make sure to add XZ Utils to the dependency list for SMServer12 & fix the build script to check for it instead of SMServer.app
- You are correct that libsmserver is 64-bit only. Since up until this point, SMServer only supported iOS 13+, there was no need to build it for armv7. I can update the makefile to add support for armv7 as well.

**Comment by vpark** [comment]
Does this mean that the next version will support iOS 12 out of the box? I tried the above method on my iOS 12.5.3 but encountered the same half built package error CSab6482 was seeing.

When I rebooted before cleaning up the package I wasn't able to rejailbreak and had to completely reset my phone.

**Comment by yeralin** [comment]
> I just fixed [the link](https://byu.box.com/v/smserver12-080)
![image](https://user-images.githubusercontent.com/8620461/152288073-fad9b2f0-0066-4a70-bbd0-f4e42c45fe9d.png)

Hey, your link is dead. Could you please upload the file to some more reliable file hosting? Maybe Google Drive?

Trying to install it on my old iPad Mini 2 with iOS 9.3.6 :)

Thank you in advance!

**Comment by CSab6482** [comment]
> Could you please upload the file to some more reliable file hosting?

Here ya go.
https://www.mediafire.com/file/yendwdxoc5ob6rx/SMServer12.deb/file

**Comment by yeralin** [comment]
Hey, sorry I didn't realize I can build it myself using `make.sh -ndl` command.

Tried installing both `.deb` and `.ipa` to no avail on iPad Mini 2 with iOS 9.3.6.

However, Remote Messages worked perfectly. Thank you so much for a reference!

Cheers :)

**Comment by juan2three** [comment]
Any update on the support? Works flawlessly via CLI but curious about the frontend to be able to adjust settings

**Comment by sabogalc** [comment]
The last commit was in 2022 and the developer now works full-time at Beeper so I think it's safe to say the project will not be updated further unless somebody forks it.
